executor/hash_table.go

```diff
@@ -258,101 +259,137 @@ func (c *hashRowContainer) ActionSpill() memory.ActionOnExceed {
-	m.entryStore.init()
-	return m
-}
-
-// Put puts the key/rowPtr pairs to the rowHashMap, multiple rowPtrs are stored in a list.
-func (m *rowHashMap) Put(hashKey uint64, rowPtr chunk.RowPtr) {
-	oldEntryAddr := m.hashTable[hashKey]
-	e := entry{
-		ptr:  rowPtr,
-		next: oldEntryAddr,
-	}
-	newEntryAddr := m.entryStore.put(e)
-	m.hashTable[hashKey] = newEntryAddr
-	m.length++
+func newUnsafeHashTable(estCount int, keepOrder bool) *unsafeHashTable {
+	ht := new(unsafeHashTable)
+	ht.hashMap = make(map[uint64]*entry, estCount)
+	ht.entryStore = newEntryStore()
+	ht.keepOrder = keepOrder
+	return ht
+}
+
+// Put puts the key/rowPtr pairs to the unsafeHashTable, multiple rowPtrs are stored in a list.
+func (ht *unsafeHashTable) Put(hashKey uint64, rowPtr chunk.RowPtr) {
+	oldEntry := ht.hashMap[hashKey]
+	newEntry := ht.entryStore.GetStore()
+	newEntry.ptr = rowPtr
+	newEntry.next = oldEntry
+	ht.hashMap[hashKey] = newEntry
+	ht.length++
 }
 
 // Get gets the values of the "key" and appends them to "values".
-func (m *rowHashMap) Get(hashKey uint64) (rowPtrs []chunk.RowPtr) {
-	entryAddr := m.hashTable[hashKey]
-	for entryAddr != nullEntryAddr {
-		e := m.entryStore.get(entryAddr)
-		entryAddr = e.next
-		rowPtrs = append(rowPtrs, e.ptr)
+func (ht *unsafeHashTable) Get(hashKey uint64) (rowPtrs []chunk.RowPtr) {
+	entryAddr := ht.hashMap[hashKey]
+	for entryAddr != nil {
+		rowPtrs = append(rowPtrs, entryAddr.ptr)
+		entryAddr = entryAddr.next
 	}
 	// Keep the order of input.
-	for i := 0; i < len(rowPtrs)/2; i++ {
-		j := len(rowPtrs) - 1 - i
-		rowPtrs[i], rowPtrs[j] = rowPtrs[j], rowPtrs[i]
+	if ht.keepOrder {
+		for i := 0; i < len(rowPtrs)/2; i++ {
+			j := len(rowPtrs) - 1 - i
+			rowPtrs[i], rowPtrs[j] = rowPtrs[j], rowPtrs[i]
+		}
 	}
 	return
 }
 
-// Len returns the number of rowPtrs in the rowHashMap, the number of keys may be less than Len
+// Len returns the number of rowPtrs in the unsafeHashTable, the number of keys may be less than Len
 // if the same key is put more than once.
-func (m *rowHashMap) Len() int { return m.length }
+func (ht *unsafeHashTable) Len() uint64 { return ht.length }
+
+// concurrentMapHashTable is a concurrent hash table built on concurrentMap
+type concurrentMapHashTable struct {
+	hashMap    concurrentMap
+	entryStore *entryStore
+	length     uint64
+}
+
+// newConcurrentMapHashTable creates a concurrentMapHashTable
+func newConcurrentMapHashTable(estCount int) *concurrentMapHashTable {
+	ht := new(concurrentMapHashTable)
+	ht.hashMap = newConcurrentMap()
+	ht.entryStore = newEntryStore()
+	ht.length = 0
+	return ht
+}
+
+// Len return the number of rowPtrs in the concurrentMapHashTable
+func (ht *concurrentMapHashTable) Len() uint64 {
+	return ht.length
+}
+
+// Put puts the key/rowPtr pairs to the concurrentMapHashTable, multiple rowPtrs are stored in a list.
+func (ht *concurrentMapHashTable) Put(hashKey uint64, rowPtr chunk.RowPtr) {
+	newEntry := ht.entryStore.GetStore()
```

Comment:
GetStore will be called in many goroutine and cause race?

Reply:
This PR just use one thread. [the next PR ]( https://github.com/pingcap/tidb/pull/16773 )avoid this.